SYS_ATL/LoopIR_scheduling.py

```diff
@@ -1203,6 +1203,46 @@ def map_e(self, e):
         return super().map_e(e)
 
 
+class _DoRearrangeDim(LoopIR_Rewrite):
+    def __init__(self, proc, alloc_stmt, dimensions):
+        assert isinstance(alloc_stmt, LoopIR.Alloc)
+
+        self.alloc_stmt = alloc_stmt
+        self.dimensions = dimensions
+
+        super().__init__(proc)
+
+        self.proc = InferEffects(self.proc).result()
+
+    def map_s(self, s):
+        # simply change the dimension
+        if s is self.alloc_stmt:
+            # construct new_hi
+            new_hi   = [s.type.hi[i] for i in self.dimensions]
+            # construct new_type
+            new_type = LoopIR.Tensor(new_hi, s.type.is_window, s.type.type)
+
+            return [LoopIR.Alloc(s.name, new_type, s.mem, None, s.srcinfo)]
+
+        # Adjust the use-site
+        if isinstance(s, LoopIR.Assign) or isinstance(s, LoopIR.Reduce):
+            if s.name is self.alloc_stmt.name:
+                # shuffle
+                new_idx = [s.idx[i] for i in self.dimensions]
+                return [type(s)(s.name, s.type, s.cast, new_idx, s.rhs, None, s.srcinfo)]
```

Comment:
Just a note: it would be really nice if we had a method to update a subset of records in an AST node... something like:

```
s.with(idx=new_idx)
```